colibre/registration.py

```diff
@@ -208,6 +208,12 @@ def register_dust(self, catalogue):
         total_dust_fraction = sum(dust_fields)
         dust_frac_error = ""
 
+        if total_dust_fraction == 0:
```

Comment:
@MatthieuSchaller `total_dust_fraction` is a vector here, so comparing it to `int` gives an error

```
Traceback (most recent call last):
  File "/home/evgenii/anaconda3/bin/swift-pipeline", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/home/evgenii/pipeline/swift-pipeline", line 227, in <module>
    catalogue = load(
  File "/home/evgenii/velociraptor-python/velociraptor/__init__.py", line 72, in load
    catalogue.register_derived_quantities(registration_file_path)
  File "/home/evgenii/velociraptor-python/velociraptor/catalogue/catalogue.py", line 413, in register_derived_quantities
    self.derived_quantities = DerivedQuantities(registration_file_path, self)
  File "/home/evgenii/velociraptor-python/velociraptor/catalogue/derived.py", line 96, in __init__
    self._register_quantities()
  File "/home/evgenii/velociraptor-python/velociraptor/catalogue/derived.py", line 111, in _register_quantities
    exec(handle.read())
  File "<string>", line 724, in <module>
  File "<string>", line 213, in register_dust
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Reply:
But it's not.  If dust is not found then you just end up with an empty list. 

This whole dust / no-dust detection is broken.

Reply:
Note that I am working on the basis of one of the recent runs you gave me:
`/cosma7/data/dp004/jlvc76/COLIBRE/cold_gas/colibre_0023.hdf5` and 
`/cosma7/data/dp004/jlvc76/COLIBRE/cold_gas/haloes.properties.0`

Reply:
@MatthieuSchaller I tried to reproduce your error by running the pipeline on the snapshot in `/cosma7/data/dp004/jlvc76/COLIBRE/cold_gas/colibre_0023.hdf5`

but got a different error

```
Traceback (most recent call last):
  File "/cosma/local/Python/3.6.5/bin/swift-pipeline", line 172, in <module>
    for input, snapshot in zip(args.input, args.snapshots)
  File "/cosma/local/Python/3.6.5/bin/swift-pipeline", line 172, in <listcomp>
    for input, snapshot in zip(args.input, args.snapshots)
  File "/cosma/home/dp004/dc-chai1/.local/lib/python3.6/site-packages/swiftsimio/__init__.py", line 95, in load
    return SWIFTDataset(filename, mask=mask)
  File "/cosma/home/dp004/dc-chai1/.local/lib/python3.6/site-packages/swiftsimio/reader.py", line 1301, in __init__
    self.get_metadata()
  File "/cosma/home/dp004/dc-chai1/.local/lib/python3.6/site-packages/swiftsimio/reader.py", line 1337, in get_metadata
    self.metadata = SWIFTMetadata(self.filename, self.units)
  File "/cosma/home/dp004/dc-chai1/.local/lib/python3.6/site-packages/swiftsimio/reader.py", line 173, in __init__
    self.load_particle_types()
  File "/cosma/home/dp004/dc-chai1/.local/lib/python3.6/site-packages/swiftsimio/reader.py", line 395, in load_particle_types
    scale_factor=self.scale_factor,
  File "/cosma/home/dp004/dc-chai1/.local/lib/python3.6/site-packages/swiftsimio/reader.py", line 634, in __init__
    self.load_metadata()
  File "/cosma/home/dp004/dc-chai1/.local/lib/python3.6/site-packages/swiftsimio/reader.py", line 654, in load_metadata
    self.load_field_descriptions()
  File "/cosma/home/dp004/dc-chai1/.local/lib/python3.6/site-packages/swiftsimio/reader.py", line 739, in load_field_descriptions
    self.field_descriptions = [get_desc(handle[x]) for x in self.field_paths]
  File "/cosma/home/dp004/dc-chai1/.local/lib/python3.6/site-packages/swiftsimio/reader.py", line 739, in <listcomp>
    self.field_descriptions = [get_desc(handle[x]) for x in self.field_paths]
  File "/cosma/home/dp004/dc-chai1/.local/lib/python3.6/site-packages/swiftsimio/reader.py", line 731, in get_desc
    description = dataset.attrs["Description"].decode("utf-8")
AttributeError: 'str' object has no attribute 'decode'
```

Is our snapshot just a copy of one of the hawk runs' snapshots?

Reply:
Try `/cosma7/data/dp004/jlvc76/COLIBRE/cold_gas/colibre_copy_0023.hdf5` sorry. The other one had surgery done to it. The one with copy in the same is straight taken from you. I think it's the one you were using to try to debug the restarting from snapshot.

Reply:
In my original halo.properties file (in `/cosma7/data/dp004/dc-chai1/HAWK/01_reference_L006N0188`), I have the fields

```
├BirthDensities_max_star  [float64: 5898] (5 attributes)
├BirthDensities_min_star  [float64: 5898] (5 attributes)
├Configuration (159 attributes)
├DensitiesAtLastSupernovaEvent_max_gas    [float64: 5898] (5 attributes)
├DustMassFractions_index_0_averagemassweighted_gas        [float64: 5898] (5 attributes)
├DustMassFractions_index_1_averagemassweighted_gas        [float64: 5898] (5 attributes)
├Efrac    [float64: 5898] (4 attributes)
├Efrac_gas        [float64: 5898] (4 attributes)
├Efrac_star       [float64: 5898] (4 attributes)
├Ekin     [float64: 5898] (4 attributes)
├ElementMassFractions_index_0_averagemassweighted_gas     [float64: 5898] (5 attributes)
├ElementMassFractions_index_4_averagemassweighted_gas     [float64: 5898] (5 attributes)
├Epot     [float64: 5898] (4 attributes)
├File_id  [int32: 1]
├ID       [uint64: 5898] (4 attributes)

```

Your catalogue file (`/cosma7/data/dp004/jlvc76/COLIBRE/cold_gas/haloes.properties.0`) has fewer fields

```
├BirthDensities_max_star  [float64: 6101] (5 attributes)
├BirthDensities_min_star  [float64: 6101] (5 attributes)
├Configuration (162 attributes)
├Efrac    [float64: 6101] (4 attributes)
├Efrac_gas        [float64: 6101] (4 attributes)
├Efrac_star       [float64: 6101] (4 attributes)
├Ekin     [float64: 6101] (4 attributes)
├Epot     [float64: 6101] (4 attributes)
├File_id  [int32: 1]
├ID       [uint64: 6101] (4 attributes)
```

I.e. the **element mass fraction** field is missing, for example. That is why the pipeline crashes with the error you reported.

Reply:
The disappearance of the additional gas fields might be related to the missing trailing commas in the` internal gas properties` section in the newest VR config. I add a comment on GitLab indicating the exact lines

Reply:
> But it's not. If dust is not found then you just end up with an empty list.
> 
> This whole dust / no-dust detection is broken.

Sorry, I don't understand. If there is no dust, shouldn't it err with the `AttrubuteError`, handle this error with the exception branch, and continue running?

Reply:
Oh, yes you are right. It is written in such a way it will never get to `getattr()` call if there is no dust. I didn't realise it. I will rewrite the code.